sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIO.java

```diff
@@ -711,17 +751,23 @@ public String asPath() {
           getSubscriptionProvider() == null
               ? null
               : NestedValueProvider.of(getSubscriptionProvider(), new SubscriptionPathTranslator());
-      PubsubUnboundedSource<T> source =
-          new PubsubUnboundedSource<T>(
+      PubsubUnboundedSource source =
+          new PubsubUnboundedSource(
               FACTORY,
               projectPath,
               topicPath,
               subscriptionPath,
-              getCoder(),
               getTimestampAttribute(),
               getIdAttribute(),
-              getParseFn());
-      return input.getPipeline().apply(source);
+              getNeedsAttributes());
+      return input
+          .getPipeline()
+          .apply(source)
+          .setCoder(
+              getNeedsAttributes()
+                  ? PubsubMessageWithAttributesCoder.of()
+                  : PubsubMessagePayloadOnlyCoder.of())
+          .apply(MapElements.via(getParseFn()));
```

Comment:
MapElements is needed regardless: the current PR makes PubsubUnboundedSource always produce PubsubMessage (with or without attributes; giving the native Dataflow source a fake identity "parse fn"); so we need to wrap it with parsing here.

I.e. we replace parsing that happened natively in PubsubUnboundedSource, with parsing here using MapElements.